File: tsa_java/app/src/test/java/tsa_java/ApartmentAllocationTest.java
package tsa_java;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

import java.util.Arrays;

public class ApartmentAllocationTest {

    @Test
    void testExactMatch() {
        int[] A = {50};
        int[] B = {50};
        int N = 1, M = 1, K = 0;
        assertEquals(1, ApartementAllocation.solve(A, B, N, M, K), "Test when applicant demand and apartment size are exactly equal.");
    }

    @Test
    void testSingleNoMatch() {
        int[] A = {50};
        int[] B = {60};
        int N = 1, M = 1, K = 5;
        assertEquals(0, ApartementAllocation.solve(A, B, N, M, K), "Test when a single applicant cannot match any apartment due to tight tolerance.");
    }

    @Test
    void testSingleMatchWithinTolerance() {
        int[] A = {50};
        int[] B = {53};
        int N = 1, M = 1, K = 3;
        assertEquals(1, ApartementAllocation.solve(A, B, N, M, K), "Test when a single applicant can match within tolerance.");
    }

    @Test
    void testMultipleExactMatch() {
        int[] A = {40, 50, 60};
        int[] B = {40, 50, 60};
        int N = 3, M = 3, K = 0;
        assertEquals(3, ApartementAllocation.solve(A, B, N, M, K), "Test when multiple applicants match exactly with apartments.");
    }

    @Test
    void testMultipleNoMatch() {
        int[] A = {10, 20, 30};
        int[] B = {100, 200, 300};
        int N = 3, M = 3, K = 10;
        assertEquals(0, ApartementAllocation.solve(A, B, N, M, K), "Test when no applicant matches any apartment due to large differences.");
    }

    @Test
    void testSomeMatches() {
        int[] A = {10, 30, 50, 70};
        int[] B = {20, 35, 60, 80};
        int N = 4, M = 4, K = 10;
        assertEquals(4, ApartementAllocation.solve(A, B, N, M, K), "Test when some applicants match with apartments and others do not.");
    }

    @Test
    void testAllApartmentsMatchMultipleApplicants() {
        int[] A = {40, 50, 60};
        int[] B = {55};
        int N = 3, M = 1, K = 10;
        assertEquals(1, ApartementAllocation.solve(A, B, N, M, K), "Test when a single apartment matches one of many applicants.");
    }

    @Test
    void testAllApplicantsMatchMultipleApartments() {
        int[] A = {50};
        int[] B = {40, 50, 60};
        int N = 1, M = 3, K = 10;
        assertEquals(1, ApartementAllocation.solve(A, B, N, M, K), "Test when a single applicant matches one of many apartments.");
    }

    @Test
    void testNoApplicants() {
        int[] A = {};
        int[] B = {40, 50, 60};
        int N = 0, M = 3, K = 10;
        assertEquals(0, ApartementAllocation.solve(A, B, N, M, K), "Test when there are no applicants.");
    }

    @Test
    void testNoApartments() {
        int[] A = {40, 50, 60};
        int[] B = {};
        int N = 3, M = 0, K = 10;
        assertEquals(0, ApartementAllocation.solve(A, B, N, M, K), "Test when there are no apartments.");
    }

    @Test
    void testZeroTolerance() {
        int[] A = {50, 51, 52};
        int[] B = {50, 51, 52};
        int N = 3, M = 3, K = 0;
        assertEquals(3, ApartementAllocation.solve(A, B, N, M, K), "Test when tolerance is zero but demands match exactly.");
    }

    @Test
    void testLargeTolerance() {
        int[] A = {10, 20, 30};
        int[] B = {100, 200, 300};
        int N = 3, M = 3, K = 1000;
        assertEquals(3, ApartementAllocation.solve(A, B, N, M, K), "Test when tolerance is very large and all match.");
    }

    @Test
    void testDuplicatesInApplicants() {
        int[] A = {50, 50, 50};
        int[] B = {50};
        int N = 3, M = 1, K = 0;
        assertEquals(1, ApartementAllocation.solve(A, B, N, M, K), "Test when applicants have duplicate demands.");
    }

    @Test
    void testDuplicatesInApartments() {
        int[] A = {50};
        int[] B = {50, 50, 50};
        int N = 1, M = 3, K = 0;
        assertEquals(1, ApartementAllocation.solve(A, B, N, M, K), "Test when apartments have duplicate sizes.");
    }

    @Test
    void testAscendingOrder() {
        int[] A = {10, 20, 30};
        int[] B = {15, 25, 35};
        int N = 3, M = 3, K = 5;
        assertEquals(3, ApartementAllocation.solve(A, B, N, M, K), "Test with already sorted arrays.");
    }

    @Test
    void testDescendingOrder() {
        int[] A = {30, 20, 10};
        int[] B = {35, 25, 15};
        int N = 3, M = 3, K = 5;
        assertEquals(3, ApartementAllocation.solve(A, B, N, M, K), "Test with arrays sorted in descending order.");
    }

    @Test
    void testMixedOrder() {
        int[] A = {20, 30, 10};
        int[] B = {25, 15, 35};
        int N = 3, M = 3, K = 5;
        assertEquals(3, ApartementAllocation.solve(A, B, N, M, K), "Test with unsorted arrays.");
    }

    @Test
    void testLargeInput() {
        int N = 1000, M = 1000, K = 10;
        int[] A = new int[N];
        int[] B = new int[M];
        Arrays.fill(A, 50);
        Arrays.fill(B, 55);
        assertEquals(1000, ApartementAllocation.solve(A, B, N, M, K), "Test with large input arrays where all match.");
    }

    @Test
    void testSparseMatches() {
        int[] A = {10, 30, 50, 70};
        int[] B = {15, 35, 55, 85};
        int N = 4, M = 4, K = 2;
        assertEquals(0, ApartementAllocation.solve(A, B, N, M, K), "Test when sparse matches occur due to small tolerance.");
    }

    @Test
    void testEdgeCaseZeroToleranceNoMatch() {
        int[] A = {10, 20, 30};
        int[] B = {15, 25, 35};
        int N = 3, M = 3, K = 0;
        assertEquals(0, ApartementAllocation.solve(A, B, N, M, K), "Test edge case with zero tolerance and no matches.");
    }
}

File: tsa_dart/test/apartment_allocation_test.dart
import 'package:test/test.dart';
import '../lib/apartment_allocation.dart';

void main() {
  group('Apartment Allotment Tests', () {
    test('Basic functionality with small arrays', () {
      expect(solve([60, 45], [30, 60], 2, 2, 5), 1);
    });

    test('Exact match for all applicants and apartments', () {
      expect(solve([40, 50, 60], [40, 50, 60], 3, 3, 0), 3);
    });

    test('No possible matches', () {
      expect(solve([10, 20, 30], [40, 50, 60], 3, 3, 5), 0);
    });

    test('All applicants have larger demands than apartments', () {
      expect(solve([70, 80, 90], [10, 20, 30], 3, 3, 10), 0);
    });

    test('All apartments are smaller but within tolerance', () {
      expect(solve([70, 80, 90], [60, 70, 80], 3, 3, 10), 3);
    });

    test('Single applicant and apartment, within tolerance', () {
      expect(solve([50], [55], 1, 1, 5), 1);
    });

    test('Single applicant and apartment, out of tolerance', () {
      expect(solve([50], [70], 1, 1, 5), 0);
    });

    test('Large arrays with no matches', () {
      expect(solve(
          List.generate(1000, (i) => i * 2),
          List.generate(1000, (i) => i * 2 + 1),
          1000, 1000, 0), 0);
    });

    test('Large arrays with all matches', () {
      expect(solve(
          List.generate(1000, (i) => i * 2),
          List.generate(1000, (i) => i * 2),
          1000, 1000, 0), 1000);
    });

    test('Random numbers, some matches within tolerance', () {
      expect(solve([30, 40, 60], [35, 45, 50], 3, 3, 10), 3);
    });

    test('Unequal array sizes, fewer applicants', () {
      expect(solve([30, 40], [35, 45, 50], 2, 3, 10), 2);
    });

    test('Unequal array sizes, fewer apartments', () {
      expect(solve([30, 40, 50], [35, 45], 3, 2, 10), 2);
    });

    test('Tolerance is zero', () {
      expect(solve([30, 40, 50], [30, 40, 50], 3, 3, 0), 3);
    });

    test('Tolerance is very high', () {
      expect(solve([10, 20, 30], [70, 80, 90], 3, 3, 100), 3);
    });

    test('Duplicates in applicant and apartment arrays', () {
      expect(solve([30, 30, 40], [30, 30, 40], 3, 3, 0), 3);
    });

    test('Single match in a larger dataset', () {
      expect(solve([10, 20, 30, 40], [100, 200, 30], 4, 3, 0), 1);
    });

    test('Empty applicant array', () {
      expect(solve([], [10, 20, 30], 0, 3, 5), 0);
    });

    test('Empty apartment array', () {
      expect(solve([10, 20, 30], [], 3, 0, 5), 0);
    });

    test('Both arrays empty', () {
      expect(solve([], [], 0, 0, 5), 0);
    });

    test('Negative numbers with positive tolerance', () {
      expect(solve([-10, -20, -30], [-15, -25, -35], 3, 3, 5), 3);
    });
  });
}
